typing/typecore.ml

```diff
@@ -2709,12 +2709,18 @@ and type_expect_
       let (args, ty_res) = type_application env funct sargs in
       end_def ();
       unify_var env (newvar()) funct.exp_type;
-      rue {
-        exp_desc = Texp_apply(funct, args);
-        exp_loc = loc; exp_extra = [];
-        exp_type = ty_res;
-        exp_attributes = sexp.pexp_attributes;
-        exp_env = env }
+      let exp =
+        { exp_desc = Texp_apply(funct, args);
+          exp_loc = loc; exp_extra = [];
+          exp_type = ty_res;
+          exp_attributes = sexp.pexp_attributes;
+          exp_env = env } in
+      begin
+        try rue exp
+        with Error (_, _, Expr_type_clash _) as err ->
+          check_partial_application false exp;
+          raise err
```

Comment:
I agree that the patch seems correct, but I worry (slightly) about trashing of backtraces that may occur during `check_partial_application` (it uses Format, etc.). We could use the `get_raw_backtrace / raise_with_backtrace` dance to avoid this, but it makes the code much less pleasant. We could almost use `Misc.try_finally`, but unfortunately its `exceptionally` clause does not take the exception as an input, so this is not quite possible. My perfectionist self would recommend tweaking `try_finally` in this way and using it here, or defining a new Misc combinator for this.